mlir/include/mlir/Dialect/SCF/Utils/Utils.h

```diff
@@ -111,11 +111,18 @@ LogicalResult coalescePerfectlyNestedSCFForLoops(scf::ForOp op);
 void collapseParallelLoops(RewriterBase &rewriter, scf::ParallelOp loops,
                            ArrayRef<std::vector<unsigned>> combinedDimensions);
 
-/// Unrolls this for operation by the specified unroll factor. Returns failure
-/// if the loop cannot be unrolled either due to restrictions or due to invalid
-/// unroll factors. Requires positive loop bounds and step. If specified,
-/// annotates the Ops in each unrolled iteration by applying `annotateFn`.
-LogicalResult loopUnrollByFactor(
+struct UnrolledLoopInfo {
+  scf::ForOp mainLoopOp;
+  scf::ForOp epilogueLoopOp;
+};
+
+/// Unrolls this for operation by the specified unroll factor. Returns the
```

Comment:
Instead of returning nullptr values can we just make the epilogueLoop `std::optional<scf::ForOp>` ?

Reply:
Yes we can. The `mainLoopOp` can also be null when unrolling fails.

Reply:
Coming back to this, do you think `optional` should be used for both loops? Also I wonder if it is generally recommended to use `optional` instead of `null` for validity checks for milr operations.

Reply:
If both could be null, then both could be optional. I dont know if there is a consistency in MLIR on this, but I prefer optional to not having `nullptr`s floating around. `optional` makes it more clear that the value "could" be `nullopt`.

Reply:
Sounds good, switched to `std::nullopt` for both fields.